comment/src/comment.ts

```diff
@@ -0,0 +1,280 @@
+/// The line-comment token is defined on a language basis.
+export const lineCommentCmd: Command = view => {
+  return dispatchToggleComment(toggleLineComment(CommentOption.OnlyComment), view)
+}
+
+/// Uncomments the current `SelectionRange` using line-comments.
+/// The line-comment token is defined on a language basis.
+export const lineUncommentCmd: Command = view => {
+  return dispatchToggleComment(toggleLineComment(CommentOption.OnlyUncomment), view)
+}
+
+/// Comments or uncomments the current `SelectionRange` using block-comments.
+/// The block-comment tokens are defined on a language basis.
+export const toggleBlockCommentCmd: Command = view => {
+  return dispatchToggleComment(toggleBlockComment(CommentOption.Toggle), view)
+}
+
+/// Comments the current `SelectionRange` using block-comments.
+/// The block-comment tokens are defined on a language basis.
+export const blockCommentCmd: Command = view => {
+  return dispatchToggleComment(toggleBlockComment(CommentOption.OnlyComment), view)
+}
+
+/// Uncomments the current `SelectionRange` using block-comments.
+/// The block-comment tokens are defined on a language basis.
+export const blockUncommentCmd: Command = view => {
+  return dispatchToggleComment(toggleBlockComment(CommentOption.OnlyUncomment), view)
+}
+
+const dispatchToggleComment = (cmd: (st: EditorState) => Transaction | null, view: EditorView): boolean => {
+  let tr = cmd(view.state)
+  if (!tr) return false
+  view.dispatch(tr)
+  return true
+}
+
+export enum CommentOption {
+  Toggle,
+  OnlyComment,
+  OnlyUncomment,
+}
+
+export const toggleBlockComment = (option: CommentOption) => (state: EditorState): Transaction | null => {
+  type BlockCommentData = { blockComment: { open: string, close: string } | undefined } | undefined
+  const data = state.languageDataAt<BlockCommentData>("commentTokens", state.selection.primary.from)[0]
+  return data === undefined || data.blockComment === undefined || data.blockComment.open === undefined || data.blockComment.close === undefined
+    ? null
+    : new BlockCommenter(data.blockComment.open, data.blockComment.close).toggle(option, state)
+}
+
+export const toggleLineComment = (option: CommentOption) => (state: EditorState): Transaction | null => {
+  type LineCommentData = { lineComment: string | undefined } | undefined
+  const data = state.languageDataAt<LineCommentData>("commentTokens", state.selection.primary.from)[0]
+  return data === undefined || data.lineComment === undefined
+    ? null
+    : new LineCommenter(data.lineComment).toggle(option, state)
+}
+
+// This class performs toggle, comment and uncomment
+// of block comments in languages that support them.
+// The `open` and `close` arguments refer to the open and close
+// tokens of which this `BlockCommenter` is made up.
+export class BlockCommenter {
+  open: string
+  close: string
+  margin: string
+  constructor(open: string, close: string, margin: string = " ") {
```

Comment:
Not terribly important, but TypeScript allows you to abbreviate this type of class declarations with

```
class BlockCommenter {
  constructor(readonly open: string, readonly close: string, readonly margin: string = " ") {}
  // ... methods
}
```

Reply:
Good to know. I'll refactor that.